lab2/part1/myrouter_part1.py

```diff
@@ -0,0 +1,149 @@
+    #if the target IP address is in our table
+    #   pkt         Arp packet to process
+    def arp_request(self, pkt: Arp):
+        reply_ip  = pkt.targetprotoaddr
+        reply_mac = self.map[reply_ip].mac_addr
+
+        #Construct reply packet
+        etp = Ethernet(ethertype = EtherType.ARP,
+                       src       = reply_mac,
+                       dst       = pkt.senderhwaddr)
+
+        arp = Arp(operation         = ArpOperation.Reply,
+                  senderhwaddr      = reply_mac,
+                  senderprotoaddr   = reply_ip,
+                  targethwaddr      = pkt.senderhwaddr,
+                  targetprotoaddr   = pkt.senderprotoaddr)
+
+        return_pkt = etp + arp
+        return return_pkt
+
+    #Process an ARP reply in the provided packet.
+    #   pkt         Arp packet to process
+    def arp_reply(self, pkt: Arp):
+        self.add_mapping(pkt.senderprotoaddr, pkt.senderhwaddr)
+
+    """
+    Class:          ARPContext.ArpEntry
+    Description:    Table entry pairing a timestamp with a MAC address
+    """
+    class ARPEntry:
+        def __init__(self, mac_addr, time=time.time()):
+            self.mac_addr   = mac_addr
+            self.timestamp  = time
+            return
+
+# end class ARPContext
+
+class Router(object):
+    def __init__(self, net):
+        self.net = net
+        self.local_proto_eth = ARPContext()     #Local address maps
+        self.other_proto_eth = ARPContext()     #Other address maps
+
+        #Cache IP->MAC mapping for local interfaces
+        for intf in net.interfaces():
+            self.local_proto_eth[intf.ipaddr] = intf.ethaddr
+        
+    #Main Router loop
+    def router_main(self):    
+        '''
+        Main method for router; we stay in a loop in this method, receiving
+        packets until the end of time.
+        '''
+        while True:
+            gotpkt = True
+            try:
+                _, input_port, pkt = self.net.recv_packet(timeout=1.0)
+            except NoPackets:
+                log_debug("No packets available in recv_packet")
+                gotpkt = False
+            except Shutdown:
+                log_debug("Got shutdown signal")
+                break
+
+            if gotpkt:
+                log_debug("Got a packet: {}".format(str(pkt)))
+
+                #Handle ARP
+                self.handle_arp(pkt.get_header(Arp), input_port)
+
+                #TODO: Handle other packets
+
+    #ARP Handler
+    def handle_arp(self, pkt: Arp, input_port: EthAddr):
+
+        if isinstance(pkt, Arp) and pkt.targetprotoaddr in self.local_proto_eth:
+            if pkt.operation == ArpOperation.Request:
+                self.send_packet(self.local_proto_eth.arp_request(pkt), input_port)
+            
+            elif pkt.operation == ArpOperation.Reply:
+                self.other_proto_eth.arp_reply(pkt)
+
+    #Packet sender
+    def send_packet(self, pkt, output_port):
+        if not isinstance(pkt, Packet):return
```

Comment:
Since a type check on `pkt` is performed, should a type check on `output_port` also be performed?

Rather than silently handling the wrongly typed `pkt`, an exception should be raised, or a `log_debug()` be included, or both.

Reply:
Added log debug output.